### PR TITLE
[deb_x64] Build go1.16 from source

### PR DESCRIPTION
**Pull request by KSerrania**
### What does this PR do?

Builds `go1.16` from source in the `deb_x64` builder, using `gcc 4.9` and `go1.15.11`.

### Motivation

With the publicly distributed `go1.16` binaries, some weird linker errors appear. For instance, `inv -e install-tools` fails with:
```
/root/.gimme/versions/go1.16.7.linux.amd64/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: /tmp/go-link-101083533/000010.o(.text+0x74): unresolvable AWAVAUATUSH��h�*H�T$(������ relocation against symbol `stderr@@GLIBC_2.2.5'
/usr/bin/ld: BFD (GNU Binutils for Debian) 2.22 internal error, aborting at ../../bfd/reloc.c line 443 in bfd_get_reloc_size
/usr/bin/ld: Please report this bug.
```

According to https://github.com/golang/go/issues/43996#issuecomment-782547917, this is due to the builder having a lower `binutils` version than what was used to build the publicly distributed go binaries.

### CODE REVIEW

File: setup_go.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# in the post-compilation steps of go)
+# We don't add it to the PATH so that it's not used in the Agent / rtloader builds
+apt-get update && apt-get install -y gcc-4.9-backport
+
+git clone https://go.googlesource.com/go goroot && cd goroot
+git checkout go$GO_VERSION
+cd src 
+
+# Use gcc 4.9 + go1.15.11 to build the target go version
+CC=/usr/lib/gcc-4.9-backport/bin/gcc ./all.bash
```
Comment (mx-psi) [inline setup_go.sh:20]: `make.bash` has these two variables. If we don't set `CC_FOR_TARGET`, does that mean that we use `gcc` 4.9 for cgo?

```
# CC: Command line to run to compile C code for GOHOSTARCH.
# Default is "gcc". Also supported: "clang".
#
# CC_FOR_TARGET: Command line to run to compile C code for GOARCH.
# This is used by cgo.  Default is CC.
```

File: setup_go.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+curl -sL -o /bin/gimme https://raw.githubusercontent.com/travis-ci/gimme/master/gimme
+chmod +x /bin/gimme
+eval $(gimme 1.15.11)
+
+# Install gcc 4.9 (needed to compile go, as it introduces -fdiagnostics-color which is required
+# in the post-compilation steps of go)
+# We don't add it to the PATH so that it's not used in the Agent / rtloader builds
+apt-get update && apt-get install -y gcc-4.9-backport
+
+git clone https://go.googlesource.com/go goroot && cd goroot
```
Comment (mx-psi) [inline setup_go.sh:15]: Should we delete `goroot` (except for `/goroot/bin`) once the build is finished? (Or, alternatively, do this on a separate build stage and just copy the binary)

Comment (KSerrania) [reply setup_go.sh:15]: I can try doing this. However, `gimme` does install the equivalent of everything that's in `goroot` when you use it to install a go version (and not only the binaries), so I wonder if it has a specific reason to do so.

File: setup_go.sh (hunk 1 above)
Comment (KSerrania) [reply setup_go.sh:20]: The only thing I could find is old docs from when the `CC_FOR_TARGET` option was introduced, but I believe the nuance is that `CC_FOR_TARGET` is used by `cgo` _when cross-compiling_: https://github.com/golang/go/commit/88d07b2cbcf0ad5e2e9936bcaec4fd25e38bec10, whereas `CC` is used for `cgo` when not cross-compiling (as the `for GOHOSTARCH` / `for GOARCH` distinction seems to show).

File: setup_go.sh (hunk 2 above)
Comment (mx-psi) [reply setup_go.sh:15]: It looks like the official Go tar.gz for Linux includes the source code and some more stuff (e.g. IANA timezone database info), so it's probably better to keep it. Marking this as solved.

File: setup_go.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+git checkout go$GO_VERSION
```
Comment (mx-psi) [inline setup_go.sh:16]:
```suggestion
git clone --branch go$GO_VERSION --depth 1 https://go.googlesource.com/go goroot && cd goroot
```

File: setup_go.sh (hunk 1 above)
Comment (mx-psi) [reply setup_go.sh:20]: I see. So if I understood correctly, the Go compiler uses cgo and these variables refer to the C compiler that you use for cgo when compiling/cross-compiling the Go compiler, but not to the one used by the Go compiler, which is independent, right?

Comment (KSerrania) [reply setup_go.sh:20]: Yes, I believe that's it. To verify that this is correct, I'll modify the script to remove the `gcc 4.9` compiler after building go.

Comment (KSerrania) [reply setup_go.sh:20]: Yes, I believe that's it. To verify that this is correct, I'll modify the script to remove the `gcc 4.9` compiler after building go.

File: setup_go.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+set -ex
+
+# Install gimme to get go1.15.11
+curl -sL -o /bin/gimme https://raw.githubusercontent.com/travis-ci/gimme/master/gimme
+chmod +x /bin/gimme
+eval $(gimme 1.15.11)
```
Comment (sgnn7) [inline setup_go.sh:8]:
```suggestion
eval "$(gimme 1.15.11)"
```

File: setup_go.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+set -ex
+
+# Install gimme to get go1.15.11
+curl -sL -o /bin/gimme https://raw.githubusercontent.com/travis-ci/gimme/master/gimme
```
Comment (sgnn7) [inline setup_go.sh:6]: Is there any way to use a stable version with a hash check of this file?

File: setup_go.sh [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+chmod +x /bin/gimme
+eval $(gimme 1.15.11)
+
+# Install gcc 4.9 (needed to compile go, as it introduces -fdiagnostics-color which is required
+# in the post-compilation steps of go)
+# We don't add it to the PATH so that it's not used in the Agent / rtloader builds
+apt-get update && apt-get install -y gcc-4.9-backport
+
+git clone --branch go$GO_VERSION --depth 1 https://go.googlesource.com/go goroot && cd goroot
```
Comment (sgnn7) [inline setup_go.sh:15]:
```suggestion
git clone --branch "go$GO_VERSION" --depth 1 https://go.googlesource.com/go goroot && cd goroot
```

File: setup_go.sh [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+cd src
```
Comment (sgnn7) [inline setup_go.sh:16]: I'm a bit curious why `cd goroot` is inlined above but we cd again here on a separate line. Probably should join those two together.

File: setup_go.sh [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+# Remove gcc 4.9 after building go
+apt-get remove -y gcc-4.9-backport
+
+# HACK: cgo tries to look for gcc in the same place that CC
+# pointed to when go was compiled. By default, this is fine (since CC=gcc),
+# but here since CC is more specific, cgo will try to find gcc
+# in /usr/lib/gcc-4.9-backport/bin/gcc
+mkdir -p /usr/lib/gcc-4.9-backport/bin
+ln -s $(which gcc) /usr/lib/gcc-4.9-backport/bin/gcc
```
Comment (sgnn7) [inline setup_go.sh:29]:
```suggestion
ln -s "$(which gcc)" /usr/lib/gcc-4.9-backport/bin/gcc
```

**Review by sgnn7 — COMMENTED** [review]
Left just a few tiny nit comments

File: setup_go.sh (hunk 5 above)
Comment (KSerrania) [reply setup_go.sh:6]: I can add a check for this (I copypasted what was in the `Dockerfile` of the builder before).

File: setup_go.sh (hunk 7 above)
Comment (KSerrania) [reply setup_go.sh:16]: Yeah, it comes from a previous version of this script where the clone + checkout were done in 2 steps.

**Review by mx-psi — APPROVED** [review]
LGTM

File: setup_go.sh [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+# Use gcc 4.9 + go1.15.11 to build the target go version
+# HACK: cgo tries to look for gcc in the same place that CC
+# pointed to when go was compiled. By default, CC=gcc.
+# We don't want to keep gcc 4.9 after building go, so we temporarily
+# change the PATH so that "gcc" is the gcc 4.9 we installed.
+# Then, when cgo is used, it will use "gcc" which will be gcc 4.7.
+export PATH="/usr/lib/gcc-4.9-backport/bin:$PATH"
+./all.bash
+
```
Comment (djmitche) [inline setup_go.sh:33]: It [looks like](https://github.com/golang/go/blob/181e8cde301cd8205489e746334174fee7290c9b/src/make.bash#L212-L213) this "bootstraps" by building the new version with whatever old version it is given (1.15.11 in this case) and then rebuilding the new version with the newly-built binary.  So, it shouldn't matter at all what the old version is, as it only needs to be able to build a working compiler.

**Review by djmitche — APPROVED** [review]

**Pull request merged**